react-components/src/components/Minicart/cartInitializer.js

```diff
@@ -47,24 +47,44 @@ const CartInitializer = props => {
             addItem: ev => {
                 if (!ev.detail) return;
 
-                let cartItems = ev.detail.items.map(item => {
+                const mapper = item => {
                     return {
                         data: {
                             sku: item.sku,
                             quantity: item.quantity
                         }
                     };
-                });
+                };
+
+                let physicalCartItems = ev.detail.filter(item => !item.virtual).map(mapper);
+                let virtualCartItems = ev.detail.filter(item => item.virtual).map(mapper);
+
                 dispatch({ type: 'open' });
                 dispatch({ type: 'beginLoading' });
 
-                let addItemFunc = ev.detail.virtual ? addVirtualItem : addItem;
+                let promises = [];
 
-                return addItemFunc({
-                    variables: { cartId, cartItems },
-                    refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
-                    awaitRefetchQueries: true
-                })
+                if (physicalCartItems.length > 0) {
+                    promises.push(
+                        addItem({
+                            variables: { cartId, cartItems: physicalCartItems },
+                            refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
+                            awaitRefetchQueries: true
+                        })
+                    );
+                }
+
+                if (virtualCartItems.length > 0) {
+                    promises.push(
+                        addVirtualItem({
+                            variables: { cartId, cartItems: virtualCartItems },
+                            refetchQueries: [{ query: CART_DETAILS_QUERY, variables: { cartId } }],
+                            awaitRefetchQueries: true
+                        })
+                    );
+                }
+
+                return Promise.all(promises)
```

Comment:
I have two comments here:
1. Why don't we do the two mutations at the same time, I mean in the same GraphQL request?
2. Whether we do 1. or not, and like with `Promise.all`, don't we have here a typical case of possible "lost update" because of the concurrent calls? Depending on the ordering and/or concurrency, it could well be that the last CART_DETAILS_QUERY would miss one add-to-cart and would only show the result of one request.

I don't know if there are any guidelines here in the Magento documentation to better understand the possible race condition.

Reply:
I discussed with @dplaton and there are basically two ways of approaching this problem. We can define a third query which combines both mutations like

```graphql
mutation addBundleToCart( params ) {
  addVirtualProductsToCart( params ) {
    cart {
      id
    }
  }
  addSimpleProductsToCart( params ) {
    cart {
      id
    }
  }
}
```

and use it in case we have both simple and virtual products. This might become problematic, if we need to support additional product types (e.g. bundled products?). Alternatively, we could let Apollo automatically batch requests https://blog.apollographql.com/batching-client-graphql-queries-a685f5bcd41b, but this solution is actually not recommended.

I couldn't observe a race condition as of now, since there seem to be some internal optimisations in Apollo that make sure the cart refetch is done only once after both mutations. But I can adapt the code to make sure.

Reply:
It's up to you, if you checked that the refetch is only done once at the right time, then it's fine. I thought there would maybe be an option in the Apollo client to "merge" multiple GraphQL mutations in one single request.

Reply:
The batching link has this part:

> This isn’t suggesting to write one large GraphQL query at the container-level. Instead, write queries normally, next to the components that use them. When you’re ready to optimize a section of an app, convert those queries to fragments and export them from the component file. You can then import these fragments in the container, let the container make the single, large query, and pass the fragment results back to the children.

So maybe instead of writing all possible combinations, we could use fragments and merge them as needed?

Reply:
I combined both queries now which also makes sure that the refetching is done only once. So no more race conditions.